File: python/algorithms/sort/insertion/iterative.py
import sys

def insert(toSort, dropIndex):
    previous = dropIndex - 1
    key = toSort[dropIndex]

    while previous > -1 and key < toSort[previous]:
        # Shift right.
        toSort[dropIndex] = toSort[previous]

        dropIndex -= 1
        previous -= 1

    toSort[dropIndex] = key

def insertionSort(toSort):
    for i in range(1, len(toSort)):
        insert(toSort, i)

    return toSort

def main(argv):
    toSort = [188, 9, 7, 1, 5, 7777, 4, 3, 1, -99]
    print('initial sort order', toSort)
    print('            sorted', insertionSort(toSort))

if __name__ == '__main__':
    main(sys.argv[1:])

File: c/algorithms/sort/insertion/iterative.c
#include "../sortlib.h"

int array[10] = { 188, 9, 7, 1, 5, 7777, 4, 3, 1, -99 };
unsigned int short len = sizeof(array) / sizeof(int);

void insert(unsigned int short dropIndex) {
    int short previous = dropIndex - 1;
    int key = array[dropIndex];

    while (previous > -1 && key < array[previous]) {
        // Shift right.
        array[dropIndex] = array[previous];

        dropIndex -= 1;
        previous -= 1;
    }

    array[dropIndex] = key;
}

void insertionSort() {
    unsigned int short i;

    for (i = 1; i < len; i++) {
        insert(i);
    }
}

void main(void) {
    printArray("Before sort:", array, len);
    insertionSort();
    printArray("After sort:", array, len);
}

